### ISSUE TITLE
_pywrap_tensorflow_internal.pyd as 'BINARY' and 'EXTENSION'

### ISSUE THREAD

**Issue by wdixon** [pyinstaller/pyinstaller#5236]
## Duplicate _pywrap_tensorflow_internal.pyd files

Trying to package at TF2.3 application - and finding two identical versions of the tensorflow runtime are crated in under the distribution directory (each 700+MB):

1.) the top level of the dist/run directory: _pywrap_tensorflow_internal.pyd
2.) The second under dist/run/tensorflow/python/_pywrap_tensorflow_internal.pyd

Looking a bit into the output of Analysis, they appear twice in the binaries section...  One with the 'BINARY' category, the other with the 'EXTENSION' category.

('tensorflow.python._pywrap_tensorflow_internal', 'c:\\users\\me\\appdata\\local\\programs\\python\\python38\\lib\\site-packages\\tensorflow\\python\\_pywrap_tensorflow_internal.pyd', 'EXTENSION'), 

('_pywrap_tensorflow_internal.pyd', 'c:\\users\\me\\appdata\\local\\programs\\python\\python38\\lib\\site-packages\\tensorflow\\python\\_pywrap_tensorflow_internal.pyd', 'BINARY')

## Context information (for bug reports)

* Output of `pyinstaller --version`: ```4.0```
* Version of Python: e.g. 3.8
* Platform: Windows
* Did you also try this on another platform? No

**Comment by rokm** [comment]
Sounds like pyinstaller/pyinstaller-hooks-contrib#49, which should be already taken care of by pyinstaller/pyinstaller-hooks-contrib#50. Can you try updating `pyinstaller-hooks-contrib` to `v2020.9`?

**Comment by wdixon** [comment]
This did address the duplicate _pywrap_tensorflow_internal.pyd,

Interestingly - the v2020.9 hooks - ends up producing ~100MB larger dist folder.
These new modules appear in the output:

17M     dist/run/tensorflow/compiler/tf2xla/ops
17M       dist/run/tensorflow/compiler/tf2xla
17M       dist/run/tensorflow/compiler
1.9M    dist/run/tensorflow/lite/python/optimize
76K     dist/run/tensorflow/python/framework


I am assuming I can just add these to the excludes - but have not tried this...

**Comment by wdixon** [comment]
Adding those to the excludes - causes a ImportError: cannot import name 'c_api_util' from 'tensorflow.python.framework' (unknown location)

So these packages weren't required before - but not entirely sure the proper way to exclude?

**Comment by rokm** [comment]
The updated hook is a bit more thorough at collecting files from the `tensorflow` (sub)modules. (Does the program built with pre-v.2020.9 hooks actually run? In my test cases, just importing `tensorflow` produces error about missing `tensorflow\lite\experimental\microfrontend\python\ops\_audio_microfrontend_op.so`).

The `tensorflow\python\framework\fast_tensor_util.so` seems to be optionally loaded (and I'm not actually sure if the import attempt succeeds on Windows), but you cannot exclude the whole `tensorflow.python.framework` submodule. If you check your tensorflow installation, you'll see that `tensorflow\python\framework\` contains quite a lot of python code, which is required and collected (but you don't see it directly in the dist directory).

The largest increase actually seems to come from `tensorflow\python\_pywrap_tensorflow_internal.lib`, which is very likely unnecessary.

**Comment by wdixon** [comment]
Yes - the exe ran fine before...  However as you pointed out - I did need to handle the _audo_miicrofronted_op.so:

```bash
datas=[
   (os.path.join(os.path.dirname(importlib.import_module('tensorflow').__file__),
   "lite/experimental/microfrontend/python/ops/_audio_microfrontend_op.so"),
  "tensorflow/lite/experimental/microfrontend/python/ops/")]
```

and had a filter to get rid of the duplicate pyd file - which the updated hook now handles..
```bash
excluded = [
    '_pywrap_tensorflow_internal.pyd'
    ]
a.binaries = TOC([x for x in a.binaries if x[0] not in excluded])
```

No customizations required in the spec file - but now there are extra modules indicated above...
Not entirely clear on how you would exclude them? (It ran before without those modules in the dist/run directory) - and the dist/run folder is larger than it was before (50-100MB - du rounded the output size and I don't have the exact difference to report).

**Comment by rokm** [comment]
Those are not actually modules in the regular sense (i.e., python modules or extensions) and they come up during data file collection when the hook calls `collect_data_files()`.

I will add `"**/*.lib"` pattern to the excludes for data collection as I think there's no need to collect the static/import lib files. (The `_pywrap_tensorflow_internal.lib` is 40 MB in size).

But as for the `.so` files, I think it's safer to collect them, even if some are optionally loaded, because on the other hand, some are clearly required (`_audio_microfrontend_op.so`). 

If you want to exclude them yourself, you could probably filter them out of `a.datas` similarly as you did with `'_pywrap_tensorflow_internal.pyd'` for `a.binaries`.

**Comment by bwoodsend** [comment]
Fixed is [#55](https://github.com/pyinstaller/pyinstaller-hooks-contrib/pull/55). In a months time or so you should be able to:
```shell
pip install "pyinstaller-hooks-contrib>=2020.10"
```
